tests/test_bookmark.py

```diff
@@ -1,109 +1,125 @@
 
     @staticmethod
-    def name():
-        return "tt_ga4_bookmark"
-
-
-    def streams_to_test(self):
+    def streams_to_test():
         # testing all streams creates massive quota issues
-        custom_id = self.custom_reports_names_to_ids()['Test Report 1']
         return {
             'content_group_report',
-            custom_id
+            'Test Report 1'
         }
 
+    bookmark_format = "%Y-%m-%d"
+    initial_bookmarks = {}
 
-    def manipulate_state(self, old_state):
-        manipulated_state = {
-            'bookmarks': {
-                stream_id: { os.getenv('TAP_GA4_PROPERTY_ID'): {'last_report_date': self.bookmark_date}}
-                for stream_id in old_state['bookmarks'].keys()
-            }
-        }
+    @staticmethod
+    def name():
+        return "tt_ga4_bookmark"
 
-        return manipulated_state
+    def manipulate_state(self, state: dict, new_bookmarks: dict):
+        """
+        This method will update the passed state with the new_bookmarks.
 
+        new_bookmarks must be in the format { stream: {replication_key: replication_value}}
+        """
+        new_state = deepcopy(state)
+        if new_state.get('bookmarks') is None:
+            new_state['bookmarks'] = {}
+        for stream, rep in new_bookmarks.items():
+            if new_state['bookmarks'].get(stream):
+                for value in rep.values():
+                    new_state['bookmarks'][stream][os.getenv('TAP_GA4_PROPERTY_ID')] = \
+                        {'last_report_date': value}
+            else:
+                # It is expected there will be only one key value pair.
+                # if that is not the case this will only use the last value in the key value pair
+                for value in rep.values():
+                    new_state['bookmarks'][stream] = {
+                        os.getenv('TAP_GA4_PROPERTY_ID'): {'last_report_date': value}}
+        return new_state
+
+    def get_bookmark_value(self, state, stream):
+        bookmark = state.get('bookmarks', {})
+        stream_bookmark = bookmark.get(self.get_stream_id(stream))
+        if stream_bookmark:
+            return stream_bookmark.get(os.getenv('TAP_GA4_PROPERTY_ID')).get('last_report_date')
+        return None
 
-    def streams_to_selected_fields(self):
+    @staticmethod
+    def streams_to_selected_fields():
+        # TODO - when selecting these fields I'm getting errors.
         return {
             "Test Report 1": {
                 "conversions",
-                "defaultChannelGrouping",
-                "eventName",
-                "eventCount",
-                "newUsers",
-                "enagementRate",
-                "engagedSessions",
-            },
-            'content_group_report': {
-                "date",
-                "browser",
-                "conversions",
+                # "defaultChannelGrouping",  # Are these are valid selections?
```

Comment:
Google has made a lot of field exclusion updates since this test was written. Notably `defaultChannelGrouping` is now `defaultChannelGroup`, but the whole list isn't compatible anyways. We should pick new fields, either using the google metric and dimension explorer: https://ga-dev-tools.google/ga4/dimensions-metrics-explorer/` or just using the app to pick out compatible metrics/dimensions

Reply:
Also this list should include date, since its required

Reply:
I'm not sure what good fields are but just using these does get records to pass the test, so I'm not terribly worried about it here. Also, we get the 4 automatic fields even if we don't specify them.  I updated this to remove the commented out stuff but didn't add all the automatic fields.